src/Features/Core/Portable/Snippets/RoslynLSPSnippetConverter.cs

```diff
@@ -0,0 +1,161 @@
+        /// </summary>
+        public static async Task<string> GenerateLSPSnippetAsync(Document document, int caretPosition, ImmutableArray<SnippetPlaceholder> placeholders, TextChange textChange, CancellationToken cancellationToken)
+        {
+            var extendedTextChange = await ExtendSnippetTextChangeAsync(document, textChange, placeholders, caretPosition, cancellationToken).ConfigureAwait(false);
+            return ConvertToLSPSnippetString(extendedTextChange, placeholders, caretPosition);
+        }
+
+        /// <summary>
+        /// Iterates through every index in the snippet string and determines where the
+        /// LSP formatted chunks should be inserted for each placeholder.
+        /// </summary>
+        private static string ConvertToLSPSnippetString(TextChange textChange, ImmutableArray<SnippetPlaceholder> placeholders, int caretPosition)
+        {
+            var textChangeStart = textChange.Span.Start;
+            var textChangeText = textChange.NewText;
+            Contract.ThrowIfNull(textChangeText);
+
+            using var _ = PooledStringBuilder.GetInstance(out var lspSnippetString);
+            using var disposer = PooledDictionary<int, (string identifier, int priority)>.GetInstance(out var dictionary);
+            GetMapOfSpanStartsToLSPStringItem(ref dictionary, placeholders, textChangeStart);
+
+            // Need to go through the length + 1 since caret postions occur before and after the
+            // character position.
+            // If there is a caret at the end of the line, then it's position
+            // will be equivalent to the length of the TextChange.
+            for (var i = 0; i < textChangeText.Length + 1;)
+            {
+                if (i == caretPosition - textChangeStart)
+                {
+                    lspSnippetString.Append("$0");
+
+                    // Special case for cursor position since they will occur between positions
+                    // so we still wants to insert the character following the cursor position.
+                    // Will not happen for placeholders since they have a direct mapping from position
+                    // of the identifier to their position in the TextChange text.
+                    if (i < textChangeText.Length)
+                    {
+                        lspSnippetString.Append(textChangeText[i]);
+                    }
+
+                    i++;
+                }
+
+                //Tries to see if a value exists at that position in the map, and if so it
+                // generates a string that is LSP formatted.
+                if (dictionary.TryGetValue(i, out var placeholderInfo))
+                {
+                    var str = $"${{{placeholderInfo.priority}:{placeholderInfo.identifier}}}";
+                    lspSnippetString.Append(str);
+
+                    // Skip past the entire identifier in the TextChange text
+                    i += placeholderInfo.identifier.Length;
```

Comment:
something feels very funky here.  in the above if check, you end with a ++... but then you lookup in this dictionary using that position that you mutated.  so if you have a caretposition and placeholder at the same position, it seems like this will fail.

DO you have testws for that?